server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -507,11 +511,32 @@ public static void registerRequestHandler(TransportService transportService, Sea
         TransportActionProxy.registerProxyAction(transportService, QUERY_CAN_MATCH_NODE_NAME, true, CanMatchNodeResponse::new);
     }
 
+    /**
+     * Decorates the transportService.registerRequestHandler with APM metric instrumentation.
+     */
+    public <Request extends TransportRequest> void registerRequestHandlerWithAPMMetrics(
+        String action,
+        Executor executor,
+        Writeable.Reader<Request> requestReader,
+        TransportRequestHandler<Request> handler
+    ) {
+        transportService.registerRequestHandler(action, executor, requestReader, (request, channel, task) -> {
+            var startTime = transportService.getThreadPool().relativeTimeInMillis();
+            try {
+                handler.messageReceived(request, channel, task);
+            } finally {
+                var elapsedTime = transportService.getThreadPool().relativeTimeInMillis() - startTime;
+                searchTransportMetrics.getActionLatencies().record(elapsedTime, Map.of("action", action));
```

Comment:
The map `Map.of("action", action)` passed into the parameter here is used for "attributes" (similar to "labels" or "tags" in other popular metric libraries). Attributes are the dynamic portion of the metric "key" (name) that can be used for filtering and aggregation.  They must be low cardinality.